### PR TITLE
Add a test exposing a bug with [mkdir] memoization

### PR DESCRIPTION
**Pull request by snowleopard**
@staronj and I found and diagnosed a weird Dune failure.

### CODE REVIEW

File: src/dune_engine/build_system.ml [hunk 1]
```diff
@@ -2,6 +2,10 @@ open Import
 open Memo.O
 module Action_builder = Action_builder0
 
+(* CR-someday amokhov: We should just get rid of this. It's broken in
+   file-watching mode (see [sandbox-mkdir.t]) and is also unnecessary: we should
+   just [mkdir] a directory before [chdir]ing into it because it's cheap and we
+   are already [mkdir]ing a lot -- see [Sandbox.create_dirs]. *)
```
Comment (rgrinberg) [inline src/dune_engine/build_system.ml:8]: An easy way to fix the bug would be to depend on the current run in this function. In fact, we could then use it to create sandboxed dirs as well. IMO the optimization is small but is still worth it. We cache stat calls for digests for example.

Comment (rgrinberg) [inline src/dune_engine/build_system.ml:8]: An easy way to fix the bug would be to depend on the current run in this function. In fact, we could then use it to create sandboxed dirs as well. IMO the optimization is small but is still worth it. We cache stat calls for digests for example.

Comment (snowleopard) [reply src/dune_engine/build_system.ml:8]: For sandboxed actions, there is nothing to optimise. We need to create directories every time.

For non-sandboxed actions, depending on the current run is not sufficient: an action might `chdir` to a dir and then delete it. Then another action in the same build run will fail.

I guess we can achieve what we want in a cheaper and a simpler way: `chdir` and then `mkdir` on exception.

**Comment by rgrinberg** [comment]
Merged.

**Pull request closed without merging**